### PR TITLE
[Improve][Core] Improve base on plugin name of lookup strategy

### LINKED ISSUE #7271: [Bug] [plugin-discovery] Jar found with same prefix
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.

### What happened

When the project began loading the plugin, we encountered multiple jars during the search due to custom development（eg ⬇️：plugin-mapping.properties）. I propose using a similarity algorithm to match the most appropriate jar. Please review the details, and I will submit the corresponding pr.

### SeaTunnel Version

2.3.3 When I raised this issue, I found that 2.3.5 had already avoided this by throwing IllegalArgumentException

### SeaTunnel Config

```conf
plugin-mapping.properties
---
seatunnel.source.HttpBase = connector-http
seatunnel.sink.HttpBase = connector-http
seatunnel.source.HttpJira = connector-http-jira
seatunnel.sink.HttpJira = connector-http-jira
seatunnel.source.Kafka = connector-kafka
seatunnel.sink.Kafka = connector-kafka
seatunnel.source.Kafka-Alcs = connector-kafka-alcs
seatunnel.sink.Kafka-Alcs = connector-kafka-alcs
seatunnel.source.Kafka-Blcs = connector-kafka-blcs
seatunnel.sink.Kafka-Blcs = connector-kafka-blcs
---
```

### Running Command

```shell
-
```

### Error Exception

```log
-
```

### Zeta or Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

image

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by corgy-w**
### Purpose of this pull request
issue : #7271 
The list of obtained jars is matched based on cosine similarity to find the closest match.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java [hunk 1]
```diff
@@ -427,20 +429,12 @@ public boolean accept(File pathname) {
                                                         pathname.getName(), pluginJarPrefix);
                                     }
                                 });
-        if (ArrayUtils.isEmpty(targetPluginFiles)) {
+        if (ArrayUtils.isEmpty(targetPluginFiles) || targetPluginFiles == null) {
```
Comment (Hisoka-X) [inline seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java]:
```suggestion
        if (ArrayUtils.isEmpty(targetPluginFiles)) {
```
`ArrayUtils.isEmpty` already contains null check.

File: seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java [hunk 2]
```diff
@@ -451,4 +445,105 @@ public boolean accept(File pathname) {
...
+            Set<String> words1 =
+                    new HashSet<>(Arrays.asList(textA.toLowerCase().split(splitRegrex)));
+            Set<String> words2 =
+                    new HashSet<>(Arrays.asList(textB.toLowerCase().split(splitRegrex)));
+            int[] termFrequency1 = calculateTermFrequencyVector(textA, words1, splitRegrex);
+            int[] termFrequency2 = calculateTermFrequencyVector(textB, words2, splitRegrex);
+            return calculateCosineSimilarity(termFrequency1, termFrequency2);
+        }
+
+        public static int[] calculateTermFrequencyVector(
```
Comment (Hisoka-X) [inline seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java]:
```suggestion
        private static int[] calculateTermFrequencyVector(
```

File: seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java [hunk 3]
```diff
@@ -427,20 +429,12 @@ public boolean accept(File pathname) {
...
-                            + Arrays.stream(targetPluginFiles)
-                                    .map(File::getPath)
-                                    .collect(Collectors.joining(","))
-                            + " for pluginIdentifier: "
-                            + pluginIdentifier);
-        }
         try {
-            URL pluginJarPath = targetPluginFiles[0].toURI().toURL();
+            URL pluginJarPath =
+                    findMostSimlarPluginJarFile(targetPluginFiles, pluginJarPrefix).toURI().toURL();
```
Comment (Hisoka-X) [inline seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java]:
```suggestion
        if (targetPluginFiles.length == 1) {
            URL pluginJarPath = targetPluginFiles[0].toURI().toURL();
        } else {
            URL pluginJarPath =
                    findMostSimlarPluginJarFile(targetPluginFiles, pluginJarPrefix).toURI().toURL();
        }
```

**Review by Hisoka-X — COMMENTED** [review]
Thanks @corgy-w ! I left some comments.

File: seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java (hunk 1 above)
Comment (corgy-w) [reply seatunnel-plugin-discovery/src/main/java/org/apache/seatunnel/plugin/discovery/AbstractPluginDiscovery.java]: yes, misled by idea editor

**Review by Hisoka-X — APPROVED** [review]
LGTM. Thanks @corgy-w !

**Comment by Carl-Zhou-CN** [comment]
good job @corgy-w

**Review by Carl-Zhou-CN — APPROVED** [review]
+1

**Pull request merged**